File: PAT_Basic_Level_by_java/1035 插入与归并/1035 插入与归并.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.Arrays;

/**
 * 1035 插入与归并
 *
 * 题目描述：
 * 给定原始序列与某一步排序后的中间序列，判断该中间序列是由插入排序还是归并排序
 * 产生的，并输出下一步排序的序列结果。
 *
 * 实现原理：
 * 1. 先模拟插入排序：每轮将第 i 个元素插入到已排序的 [0..i-1] 区间中，每轮
 *    结束后与给定序列比较；若匹配则判定为插入排序，再执行下一轮后输出。
 * 2. 若插入排序不匹配，则判定为归并排序。模拟自底向上的归并排序（先 1 个一组、
 *    再 2 个一组、4 个一组……），每轮归并后与给定序列比较，匹配后再执行一轮
 *    归并后输出。
 * 3. mergePass 方法按指定步长 len 对数组进行一趟归并，merge 方法合并两个
 *    有序子数组。
 *
 * 时间复杂度：插入排序模拟 O(n²)，归并排序模拟 O(n log n)
 */
public class Main {
    private static int n;
    private static int[] orig, given;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine().trim());
        orig = new int[n];
        given = new int[n];
        String[] p1 = br.readLine().trim().split("\\s+");
        String[] p2 = br.readLine().trim().split("\\s+");
        for (int i = 0; i < n; i++) {
            orig[i] = Integer.parseInt(p1[i]);
            given[i] = Integer.parseInt(p2[i]);
        }

        // 模拟插入排序，每轮后与给定序列比较
        int[] a = orig.clone();
        boolean isInsertion = false;
        for (int i = 1; i < n; i++) {
            // 将 a[i] 插入到已排序区间 [0..i-1] 中
            int t = a[i];
            int j = i - 1;
            while (j >= 0 && a[j] > t) {
                a[j + 1] = a[j];
                j--;
            }
            a[j + 1] = t;
            // 若当前序列与给定序列匹配，则判定为插入排序
            if (Arrays.equals(a, given)) {
                isInsertion = true;
                // 再执行下一轮插入，得到下一步的结果
                if (i + 1 < n) {
                    int t2 = a[i + 1];
                    int j2 = i;
                    while (j2 >= 0 && a[j2] > t2) {
                        a[j2 + 1] = a[j2];
                        j2--;
                    }
                    a[j2 + 1] = t2;
                }
                break;
            }
        }

        if (isInsertion) {
            System.out.println("Insertion Sort");
            print(a);
        } else {
            // 归并排序：自底向上模拟，步长从1开始逐次翻倍
            int[] b = orig.clone();
            int step = 1;
            while (step < n) {
                // 按当前步长执行一趟归并
                mergePass(b, step);
                if (Arrays.equals(b, given)) {
                    // 匹配后，再执行一趟归并作为下一步输出
                    mergePass(b, 2 * step);
                    break;
                }
                step *= 2;
            }
            System.out.println("Merge Sort");
            print(b);
        }
    }

    // 一趟归并：按步长 len 将数组分成若干对子数组并两两合并
    private static void mergePass(int[] a, int len) {
        int i = 0;
        while (i + len < n) {
            int mid = i + len;                       // 右半部分的起始位置
            int right = Math.min(i + 2 * len, n);    // 右半部分的结束位置（不超过数组长度）
            merge(a, i, mid, right);
            i = right;                               // 移动到下一对子数组
        }
    }

    // 合并两个有序子数组 [l, m) 和 [m, r) 到原数组 a 中
    private static void merge(int[] a, int l, int m, int r) {
        int[] tmp = new int[r - l];  // 临时数组存放合并结果
        int i = l, j = m, k = 0;
        // 依次取两个子数组中较小的元素放入临时数组
        while (i < m && j < r) {
            tmp[k++] = (a[i] <= a[j]) ? a[i++] : a[j++];
        }
        // 将剩余元素复制到临时数组
        while (i < m) tmp[k++] = a[i++];
        while (j < r) tmp[k++] = a[j++];
        // 将临时数组写回原数组
        for (int x = 0; x < tmp.length; x++) a[l + x] = tmp[x];
    }

    private static void print(int[] a) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            if (i > 0) sb.append(' ');
            sb.append(a[i]);
        }
        System.out.println(sb.toString());
    }
}

File: PAT_Basic_Level_by_cpp/1035 插入与归并/1035 插入与归并.cpp
/*
 * 1035 插入与归并
 *
 * 【实现原理】
 * 给定原始序列和经过某排序算法若干次迭代后的中间序列，判断是插入排序还是归并排序，
 * 并输出下一步迭代的结果。
 *
 * 插入排序特征：前若干位已有序且与原始序列前若干位相同，后面位置与原始序列相同。
 * 归并排序特征：每 2^k 个元素组成的子段内有序，段边界与原始序列不同。
 *
 * 【算法思路】
 * 1. 先试探插入排序：从原始序列开始模拟插入排序每一步，
 *    每一步结束后与给定序列比较。若匹配，则再做一步插入排序后输出结果。
 * 2. 若不是插入排序，则一定是归并排序：
 *    从原始序列开始模拟自底向上的归并排序（步长 2, 4, 8, ...），
 *    每步对每段进行排序，与给定序列比较。匹配后再做一趟归并后输出。
 *
 * 【关键点】
 * - 判断排序类型时，必须从原始序列开始模拟，而非直接检查特征
 * - 插入排序匹配后需再做一步，即再插入下一个元素
 * - 归并排序是自底向上（非递归）的方式，步长每次翻倍
 * - same 函数用于比较两个 vector 是否相等
 *
 * 【复杂度分析】
 * - 时间复杂度：O(N^2 log N)，插入排序每步 O(N)，归并排序每趟 O(N log N)
 * - 空间复杂度：O(N)，需要复制原始序列进行模拟
 */

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// 判断两个 vector 是否完全相同
bool same(const vector<int>& a, const vector<int>& b) {
    for (size_t i = 0; i < a.size(); ++i)
        if (a[i] != b[i]) return false;
    return true;
}

int main() {
    int n;
    if (!(cin >> n)) return 0;
    vector<int> orig(n), given(n);
    for (int i = 0; i < n; ++i) cin >> orig[i];
    for (int i = 0; i < n; ++i) cin >> given[i];

    // ---------- 试探插入排序 ----------
    vector<int> t = orig;
    bool isInsert = false;
    for (int i = 1; i < n; ++i) {  // 依次插入第 i 个元素
        int x = t[i], j = i - 1;
        // 将 x 插入到已排序区间 [0, i-1] 的正确位置
        while (j >= 0 && t[j] > x) { t[j + 1] = t[j]; --j; }
        t[j + 1] = x;

        if (same(t, given)) {  // 与给定序列匹配
            isInsert = true;
            // 再做一步插入排序（插入下一个元素）
            if (i + 1 < n) {
                int x2 = t[i + 1], j2 = i;
                while (j2 >= 0 && t[j2] > x2) { t[j2 + 1] = t[j2]; --j2; }
                t[j2 + 1] = x2;
            }
            break;
        }
    }

    if (isInsert) {
        cout << "Insertion Sort" << endl;
        for (int i = 0; i < n; ++i) { if (i) cout << ' '; cout << t[i]; }
        cout << endl;
        return 0;
    }

    // ---------- 归并排序（自底向上） ----------
    t = orig;
    // step 为归并子段长度，从 2 开始，每次翻倍
    for (int step = 2; step / 2 <= n; step *= 2) {
        // 对每个长度为 step 的子段分别排序
        for (int i = 0; i < n; i += step) {
            int r = min(i + step, n);  // 子段右边界（开区间）
            sort(t.begin() + i, t.begin() + r);
        }
        if (same(t, given)) {  // 与给定序列匹配
            // 再做一趟归并
            step *= 2;
            for (int i = 0; i < n; i += step) {
                int r = min(i + step, n);
                sort(t.begin() + i, t.begin() + r);
            }
            break;
        }
    }

    cout << "Merge Sort" << endl;
    for (int i = 0; i < n; ++i) { if (i) cout << ' '; cout << t[i]; }
    cout << endl;
    return 0;
}